File: ImperatorToCK3Tests/ImperatorWorldTests/Countries/CountryTests.cpp
#include "gtest/gtest.h"
#include "Imperator/Countries/Country.h"
#include "Imperator/Countries/CountryFactory.h"
#include "Imperator/Provinces/Province.h"
#include <sstream>



TEST(ImperatorWorld_CountryTests, IDCanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(42, theCountry.getID());
}

TEST(ImperatorWorld_CountryTests, tagCanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\ttag=\"WTF\"";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ("WTF", theCountry.getTag());
}

TEST(ImperatorWorld_CountryTests, tagDefaultsToBlank) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_TRUE(theCountry.getTag().empty());
}

TEST(ImperatorWorld_CountryTests, nameCanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "country_name = {\n";
	input << "name=\"WTF\"\n";
	input << "}\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ("WTF", theCountry.getName());
}

TEST(ImperatorWorld_CountryTests, nameDefaultsToBlank) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_TRUE(theCountry.getName().empty());
}

TEST(ImperatorWorld_CountryTests, capitalCanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tcapital = 32\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(32, theCountry.getCapital());
}

TEST(ImperatorWorld_CountryTests, capitalDefaultsTonNullopt) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_FALSE(theCountry.getCapital());
	ASSERT_EQ(std::nullopt, theCountry.getCapital());
}

TEST(ImperatorWorld_CountryTests, currenciesDefaultToProperValues) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(0, theCountry.getCurrencies().manpower);
	ASSERT_EQ(0, theCountry.getCurrencies().gold);
	ASSERT_EQ(50, theCountry.getCurrencies().stability);
	ASSERT_EQ(0, theCountry.getCurrencies().tyranny);
	ASSERT_EQ(0, theCountry.getCurrencies().war_exhaustion);
	ASSERT_EQ(0, theCountry.getCurrencies().aggressive_expansion);
	ASSERT_EQ(0, theCountry.getCurrencies().political_influence);
	ASSERT_EQ(0, theCountry.getCurrencies().military_experience);
}

TEST(ImperatorWorld_CountryTests, currenciesCanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tcurrency_data={ manpower=1 gold=2 stability=69 tyranny=4 war_exhaustion=2 aggressive_expansion=50 political_influence=4 military_experience=1}";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(1, theCountry.getCurrencies().manpower);
	ASSERT_EQ(2, theCountry.getCurrencies().gold);
	ASSERT_EQ(69, theCountry.getCurrencies().stability);
	ASSERT_EQ(4, theCountry.getCurrencies().tyranny);
	ASSERT_EQ(2, theCountry.getCurrencies().war_exhaustion);
	ASSERT_EQ(50, theCountry.getCurrencies().aggressive_expansion);
	ASSERT_EQ(4, theCountry.getCurrencies().political_influence);
	ASSERT_EQ(1, theCountry.getCurrencies().military_experience);
}

TEST(ImperatorWorld_CountryTests, monarchCanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tmonarch=69";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(69, *theCountry.getMonarch());
}

TEST(ImperatorWorld_CountryTests, monarchDefaultsToNullopt) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_FALSE(theCountry.getMonarch());
}

TEST(ImperatorWorld_CountryTests, color1CanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tcolor = rgb { 69 4 20 }";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(69, theCountry.getColor1().value().r());
	ASSERT_EQ(4, theCountry.getColor1().value().g());
	ASSERT_EQ(20, theCountry.getColor1().value().b());
}

TEST(ImperatorWorld_CountryTests, color1DefaultsToNullopt) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_FALSE(theCountry.getColor1());
}

TEST(ImperatorWorld_CountryTests, color2CanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tcolor2 = rgb { 69 4 20 }";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(69, theCountry.getColor2().value().r());
	ASSERT_EQ(4, theCountry.getColor2().value().g());
	ASSERT_EQ(20, theCountry.getColor2().value().b());
}

TEST(ImperatorWorld_CountryTests, color2DefaultsToNullopt) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_FALSE(theCountry.getColor2());
}

TEST(ImperatorWorld_CountryTests, color3CanBeSet) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tcolor3 = rgb { 69 4 20 }";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_EQ(69, theCountry.getColor3().value().r());
	ASSERT_EQ(4, theCountry.getColor3().value().g());
	ASSERT_EQ(20, theCountry.getColor3().value().b());
}

TEST(ImperatorWorld_CountryTests, color3DefaultsToNullopt) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_FALSE(theCountry.getColor3());
}

TEST(ImperatorWorld_CountryTests, correctCountryRankIsReturned) {
	std::stringstream input;
	Imperator::Country::Factory factory;
	
	auto theCountry1 = *factory.getCountry(input, 1);

	auto theCountry2 = *factory.getCountry(input, 2);
	theCountry2.registerProvince(std::make_shared<Imperator::Province>());
	
	auto theCountry3 = *factory.getCountry(input, 3);
	for (unsigned i = 0; i < 4; ++i) theCountry3.registerProvince(std::make_shared<Imperator::Province>());

	auto theCountry4 = *factory.getCountry(input, 4);
	for (unsigned i = 0; i < 25; ++i) theCountry4.registerProvince(std::make_shared<Imperator::Province>());

	auto theCountry5 = *factory.getCountry(input, 5);
	for (unsigned i = 0; i < 200; ++i) theCountry5.registerProvince(std::make_shared<Imperator::Province>());

	auto theCountry6 = *factory.getCountry(input, 6);
	for (unsigned i = 0; i < 753; ++i) theCountry6.registerProvince(std::make_shared<Imperator::Province>());

	ASSERT_EQ(Imperator::countryRankEnum::migrantHorde, theCountry1.getCountryRank());
	ASSERT_EQ(Imperator::countryRankEnum::cityState, theCountry2.getCountryRank());
	ASSERT_EQ(Imperator::countryRankEnum::localPower, theCountry3.getCountryRank());
	ASSERT_EQ(Imperator::countryRankEnum::regionalPower, theCountry4.getCountryRank());
	ASSERT_EQ(Imperator::countryRankEnum::majorPower, theCountry5.getCountryRank());
	ASSERT_EQ(Imperator::countryRankEnum::greatPower, theCountry6.getCountryRank());
}


TEST(ImperatorWorld_CountryTests, lawsDefaultToEmpty) {
	std::stringstream input;
	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_TRUE(theCountry.getLaws().empty());
}


TEST(ImperatorWorld_CountryTests, onlyLawsForCorrectGovernmentTypeAreReturned) {
	std::stringstream input;
	input << "= {\n";
	input << "\tsuccession_law = lawA\n";
	input << "\ttribal_authority_laws = lawB\n"; // won't be returned, law is for tribals
	input << "\trepublican_mediterranean_laws = lawC\n"; // won't be returned, law is for republics
	input << "\tmonarchy_legitimacy_laws = lawD\n";
	input << "}";
	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42); // gov type is monarchy by default

	ASSERT_EQ(2, theCountry.getLaws().size());
	ASSERT_TRUE(theCountry.getLaws().contains("lawA"));
	ASSERT_FALSE(theCountry.getLaws().contains("lawB"));
	ASSERT_FALSE(theCountry.getLaws().contains("lawC"));
	ASSERT_TRUE(theCountry.getLaws().contains("lawD"));
}


TEST(ImperatorWorld_CountryTests, wrongTypeLawsAreNotSet) {
	std::stringstream input;
	input << "= {\n";
	input << "\tnonexistent_law_type_laws = lawA\n";
	input << "}";
	const auto theCountry = *Imperator::Country::Factory().getCountry(input, 42);

	ASSERT_TRUE(theCountry.getLaws().empty());
}

File: ImperatorToCK3.UnitTests/Imperator/Countries/CountryTests.cs
using commonItems;
using ImperatorToCK3.Imperator.Countries;
using ImperatorToCK3.Imperator.Provinces;
using System.Collections.Generic;
using System.IO;
using Xunit;

namespace ImperatorToCK3.UnitTests.Imperator.Countries {
	[Collection("Sequential")]
	[CollectionDefinition("Sequential", DisableParallelization = true)]
	public class CountryTests {
		[Fact]
		public void FieldsDefaultToCorrectValues() {
			var reader = new BufferedReader(string.Empty);
			var country = Country.Parse(reader, 42);
			Assert.Equal(string.Empty, country.Tag);
			Assert.Equal(string.Empty, country.HistoricalTag);
			Assert.Equal(string.Empty, country.Name);
			Assert.Null(country.Capital);
			Assert.Equal(0, country.Currencies.Manpower);
			Assert.Equal(0, country.Currencies.Gold);
			Assert.Equal(50, country.Currencies.Stability);
			Assert.Equal(0, country.Currencies.Tyranny);
			Assert.Equal(0, country.Currencies.WarExhaustion);
			Assert.Equal(0, country.Currencies.AggressiveExpansion);
			Assert.Equal(0, country.Currencies.PoliticalInfluence);
			Assert.Equal(0, country.Currencies.MilitaryExperience);
			Assert.Null(country.Monarch);
			Assert.Null(country.Color1);
			Assert.Null(country.Color2);
			Assert.Null(country.Color3);
			Assert.Empty(country.GetLaws());
			Assert.False(country.PlayerCountry);
			Assert.Null(country.Government);
			Assert.Equal(GovernmentType.monarchy, country.GovernmentType);
		}
		[Fact]
		public void FieldsCanBeSet() {
			var reader = new BufferedReader(
				"= {\n" +
				"\ttag=\"WTF\"" +
				"\thistorical=\"WTF\"" +
				"\tcountry_name = {\n" +
				"\t\tname=\"WTF\"\n" +
				"\t}\n" +
				"\tflag=\"WTF\"" +
				"\tcapital = 32\n" +
				"\tcurrency_data={ manpower=1 gold=2 stability=69 tyranny=4 war_exhaustion=2 aggressive_expansion=50 political_influence=4 military_experience=1}" +
				"\tmonarch=69" +
				"\tprimary_culture=athenian" +
				"\treligion=hellenic" +
				"\tcolor = rgb { 1 2 3 }" +
				"\tcolor2 = rgb { 4 5 6 }" +
				"\tcolor3 = rgb { 7 8 9 }" +
				"\tgovernment_key = dictatorship" +
				"}"
			);
			var country = Country.Parse(reader, 42);
			Assert.Equal((ulong)42, country.Id);
			Assert.Equal("WTF", country.Tag);
			Assert.Equal("WTF", country.HistoricalTag);
			Assert.Equal("WTF", country.Name);
			Assert.Equal("WTF", country.Flag);
			Assert.Equal((ulong)32, country.Capital);
			Assert.Equal(1, country.Currencies.Manpower);
			Assert.Equal(2, country.Currencies.Gold);
			Assert.Equal(69, country.Currencies.Stability);
			Assert.Equal(4, country.Currencies.Tyranny);
			Assert.Equal(2, country.Currencies.WarExhaustion);
			Assert.Equal(50, country.Currencies.AggressiveExpansion);
			Assert.Equal(4, country.Currencies.PoliticalInfluence);
			Assert.Equal(1, country.Currencies.MilitaryExperience);
			Assert.Null(country.Monarch); // not linked yet
			Assert.Equal("athenian", country.PrimaryCulture);
			Assert.Equal("hellenic", country.Religion);
			Assert.Equal(new Color(1, 2, 3), country.Color1);
			Assert.Equal(new Color(4, 5, 6), country.Color2);
			Assert.Equal(new Color(7, 8, 9), country.Color3);
			Assert.Equal("dictatorship", country.Government);
			Assert.Equal(GovernmentType.monarchy, country.GovernmentType);

			var countries = new CountryCollection { country };

			var monarch = ImperatorToCK3.Imperator.Characters.Character.Parse(
				new BufferedReader("{ country=42 }"),
				"69",
				null
			);
			monarch.LinkCountry(countries);
			Assert.NotNull(country.Monarch);
			Assert.Equal((ulong)69, country.Monarch.Id);
		}

		[Fact]
		public void MultipleCountriesCanShareHistoricalTag() {
			var countriesReader = new BufferedReader(
				"1 = {\n" +
				"\ttag=\"AAA\"" +
				"\thistorical=\"AAA\"" +
				"}\n" +
				"2 = {\n" +
				"\ttag=\"BBB\"" +
				"\thistorical=\"AAA\"" +
				"\torigin=1" +
				"}\n" +
				"3 = {\n" +
				"\ttag=\"CCC\"" +
				"\thistorical=\"AAA\"" +
				"\torigin=2" +
				"}\n" +
				"4 = {\n" +
				"\ttag=\"DDD\"" +
				"\thistorical=\"AAA\"" +
				"\torigin=3" +
				"}\n"
			);
			var countries = new CountryCollection(countriesReader);
			Assert.Equal("AAA", countries[1].HistoricalTag);
			Assert.Equal("AAA", countries[2].HistoricalTag);
			Assert.Equal("AAA", countries[3].HistoricalTag);
			Assert.Equal("AAA", countries[4].HistoricalTag);
		}

		[Fact]
		public void CorrectGovernmentTypeIsRecognized() {
			var config = new Configuration {
				CK3Path = "TestFiles/CK3"
			};
			var mods = new List<Mod> {
				new("cool_mod", Path.Combine(Directory.GetCurrentDirectory(), "TestFiles/documents/CK3/mod/cool_mod"))
			};
			Country.LoadGovernments(config, mods);

			var monarchyReader = new BufferedReader("government_key = super_monarchy");
			var monarchyCountry = Country.Parse(monarchyReader, 1);
			Assert.Equal(GovernmentType.monarchy, monarchyCountry.GovernmentType);

			var republicReader = new BufferedReader("government_key = aristocratic_republic");
			var republicCountry = Country.Parse(republicReader, 2);
			Assert.Equal(GovernmentType.republic, republicCountry.GovernmentType);

			var tribalReader = new BufferedReader("government_key = tribal_federation");
			var tribalCountry = Country.Parse(tribalReader, 3);
			Assert.Equal(GovernmentType.tribal, tribalCountry.GovernmentType);
		}
		[Fact]
		public void CorrectCountryRankIsReturned() {
			var reader = new BufferedReader(string.Empty);
			var country1 = Country.Parse(reader, 1);

			var country2 = Country.Parse(reader, 2);
			country2.RegisterProvince(new Province(0));

			var country3 = Country.Parse(reader, 3);
			for (ulong i = 0; i < 4; ++i) {
				country3.RegisterProvince(new Province(i));
			}

			var country4 = Country.Parse(reader, 4);
			for (ulong i = 0; i < 25; ++i) {
				country4.RegisterProvince(new Province(i));
			}

			var country5 = Country.Parse(reader, 5);
			for (ulong i = 0; i < 200; ++i) {
				country5.RegisterProvince(new Province(i));
			}

			var country6 = Country.Parse(reader, 6);
			for (ulong i = 0; i < 600; ++i) {
				country6.RegisterProvince(new Province(i));
			}

			Assert.Equal(CountryRank.migrantHorde, country1.Rank);
			Assert.Equal(CountryRank.cityState, country2.Rank);
			Assert.Equal(CountryRank.localPower, country3.Rank);
			Assert.Equal(CountryRank.regionalPower, country4.Rank);
			Assert.Equal(CountryRank.majorPower, country5.Rank);
			Assert.Equal(CountryRank.greatPower, country6.Rank);
		}

		[Fact]
		public void OnlyLawsForCorrectGovernmentTypeAreReturned() {
			var reader = new BufferedReader(
				"= {\n" +
				"\tsuccession_law = lawA\n" +
				"\ttribal_authority_laws = lawB\n" + // won't be returned, law is for tribals
				"\trepublican_mediterranean_laws = lawC\n" + // won't be returned, law is for republics
				"\tmonarchy_legitimacy_laws = lawD\n" +
				"}"
			);
			// gov type is monarchy by default
			var country = Country.Parse(reader, 42);
			Assert.Collection(country.GetLaws(),
				item => Assert.Equal("lawA", item),
				item => Assert.Equal("lawD", item)
			);
		}

		[Fact]
		public void WrongTypeLawsAreNotSet() {
			var reader = new BufferedReader(
				"= {\n" +
				"\tnonexistent_law_type_laws = lawA\n" +
				"}"
			);
			// gov type is monarchy by default
			var country = Country.Parse(reader, 42);
			Assert.Empty(country.GetLaws());
		}

		[Fact]
		public void IgnoredTokensAreSaved() {
			var reader1 = new BufferedReader("= { monarch=20 ignoredKeyword1=something ignoredKeyword2={} }");
			var reader2 = new BufferedReader("= { ignoredKeyword1=stuff ignoredKeyword3=stuff }");
			Country.IgnoredTokens.Clear();
			_ = Country.Parse(reader1, 1);
			_ = Country.Parse(reader2, 2);

			var expectedIgnoredTokens = new HashSet<string> {
				"ignoredKeyword1", "ignoredKeyword2", "ignoredKeyword3"
			};
			Assert.True(Country.IgnoredTokens.SetEquals(expectedIgnoredTokens));
		}

		[Fact]
		public void IgnoredCountryCurrenciesTokensAreSaved() {
			var reader = new BufferedReader(
				"= { currency_data={ manpower=1 innovations=0 } }"
			);
			var reader2 = new BufferedReader(
				"= { currency_data={ ignoredKeyword1=stuff ignoredKeyword2=stuff } }"
			);
			_ = Country.Parse(reader, 1);
			_ = Country.Parse(reader2, 2);

			var expectedIgnoredTokens = new HashSet<string> {
				"ignoredKeyword1", "ignoredKeyword2", "innovations"
			};
			Assert.True(CountryCurrencies.IgnoredTokens.SetEquals(expectedIgnoredTokens));
		}
	}
}
